python-package/lightgbm/basic.py

```diff
@@ -2538,14 +2538,12 @@ def set_categorical_feature(
         self : Dataset
             Dataset with set categorical features.
         """
-        if self.categorical_feature == categorical_feature:
+        if self.categorical_feature == categorical_feature or categorical_feature == 'auto':
```

Comment:
I support moving this up here to reduce the complexity of the code but how is this related to this pull request? If it's not, can you please put up a separate PR with this change?

That would also have the added benefit of removing your status as a first-time contributor to LightGBM... so you'd no longer have to wait for a maintainer to click a button to run the CI jobs when you push a commit 😊

Reply:
This change is necessary in order to make early stopping work with setting categorical features during fit. Without this change the following error is raised

```
lightgbm.basic.LightGBMError: Cannot set categorical feature after freed raw data, set free_raw_data=False when construct Dataset to avoid this.
```

I will add a test to check that early_stopping=True works together with setting categorical features in fit.

Reply:
oh! Right right, setting `categorical_feature` after the Dataset's constructed and when you don't have the raw data around any more is problematic.

Now I'm even more confused though... how does adding automatic early stopping interact with anything related to setting categorical features?

You are just enabling *automatically* something that is already possible in `lightgbm`, changes like this shouldn't be necessary.

Reply:
The thing is that in order to enable automatic early stopping, we also need to automatically create an evaluation set.

In the current implementation, the evaluation set is created by splitting the `train_set` ([python-package/lightgbm/sklearn.py#L818-831](https://github.com/ClaudioSalvatoreArcidiacono/LightGBM/blob/a5b0c603c34949f9ed7fd927d26089c485f1aaa4/python-package/lightgbm/sklearn.py#L818-831)) instead of splitting the raw data passed to fit (X, y). This has been done to reuse the existing code for creating evaluation sets which relies on the Dataset API.

I am not very fluent with the Dataset API of lightgbm, so if you could recommend better ways for splitting train and validation I would be open for changes.

Reply:
Looks like this is no longer in the diff, so I think you found another way to do this. Marking this thread resolved.